### ISSUE TITLE
[BUG] WorldSimulation enabled state is broken

### ISSUE THREAD

**Issue by jankrassnigg** [ezEngine/ezEngine#1307]
When I try to open the Corridor scene, it asserts in `ezEngineProcessDocumentContext::GetWorldBounds`:

`EZ_ASSERT_DEV(!pWorld->GetWorldSimulationEnabled(), "World simulation must be disabled to get bounds!");`

In a different project it works, but when I press play-the-game and then exit, `OnSimulationStarted` gets called on components again, even though the editor is now in non-simulation mode.

**Comment by jankrassnigg** [comment]
I assume this now happens, because `ezWorld::SetWorldSimulationEnabled` is now ALWAYS one frame delayed. Ie, when the editor disables this flag, it will still be active for one frame and thus run the simulation for everything that already got created in that frame.

Either we need a force flag to already disable the simulation immediately, or some logic to detect whether nothing has been simulated this frame and thus it can already be disabled within the same frame.

**Comment by jankrassnigg** [comment]
Additionally, we have code that does this:

![grafik](https://github.com/ezEngine/ezEngine/assets/6001174/a435dea9-09d6-4a36-baa2-b313354dc2e5)

Since the values are now out of sync, the `GetWorldSimulationEnabled` may give you an incorrect value. Ie it may return true, and thus enable the simulation later, even though it had already been set to `false` before. Or the other way round.

I'd argue `GetWorldSimulationEnabled` should return the value that was set with `SetWorldSimulationEnabled`, even if it is still waiting for the next `Update` call.

**Comment by jankrassnigg** [comment]
In a quick test, this seems to fix the issues:

![grafik](https://github.com/ezEngine/ezEngine/assets/6001174/23771d26-158b-420e-9883-9cbe571b6ca3)

**Comment by Sanakan8472** [comment]
Ran into the same issue, created a PR here: https://github.com/ezEngine/ezEngine/pull/1308 to see if CI agrees that this is the correct fix.
I think the issue with CI is that we may be caching the conversion results so that CI didn't re-convert the model during the editor tests and thus it wasn't discovered. I'm lookin into that.

**Comment by jankrassnigg** [comment]
Issue is not fixed. Just press play in any scene, and look at the log, it will say "World simulation enabled" over and over.
Thus it also recomputes certain things all the time, which for example makes the "AI" scene completely unusable.

My fix above still works. Don't know why you went with something else. The checked in fix is a hacky work-around.

I'm very much inclined to revert the entire change.

**Comment by C-Core** [comment]
Yes please revert the entire change. I will look into a proper solution once I'm back from vacation. 
@RiederAlex FYI

**Comment by RiederAlex** [comment]
I was afraid that my change could cause issues like this.
A potentially better solution would be to provide the following behavior:
  - A frame is either completely simulated, or not
  - Set/GetWorldSimulationEnabled reflects the new state immediately if not inside an ezWorld::Update() loop
  - Set/GetWorldSimulationEnabled is delayed if inside an Update loop

This all hinges on the fact if you think "A frame is either completely simulated, or not" is still worth fixing or if this causes too much troubles.

**Comment by Sanakan8472** [comment]
Ok creating a PR to revert the changes for now: https://github.com/ezEngine/ezEngine/pull/1311
@jankrassnigg your original fix would have broken the unit tests as that was the entire intention behind the change (that only simulating can switch the active state). So I thought that I can just pin the simulation state by calling update again on any change to the state. But it seems this requires some more thought to be put into it. The editor has a very different simulation behaviour and will pause a lot and then step one frame on demand for example which causes a lot of desyncs in this new behaviour.

**Comment by Sanakan8472** [comment]
@RiederAlex I haven't dived too deeply into this code yet but in general I would prefer:
* A frame is either completely simulated, or not (agree, this should be the case)
* Any Set/Get functions of the same name always immediately reflect the change at any point in the code. Regardless of whether we are in the update loop or not.
* Ideally, any code that is run during the update loop and wants to know whether this is a simulated frame should call a different API, like `IsWorldSimulating` which is set accordingly at the start of  the `Update` function.
This would require that the parts of the code that update things are always executed as part of the `Update` function and not called from somewhere else for some other reason.

**Comment by jankrassnigg** [comment]
Not an issue anymore since the code was reverted. Currently no plans to look into this further.
